src/auth/executableresponse.ts

```diff
@@ -0,0 +1,195 @@
+// limitations under the License.
+
+const SAML_SUBJECT_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:saml2';
+const OIDC_SUBJECT_TOKEN_TYPE1 = 'urn:ietf:params:oauth:token-type:id_token';
+const OIDC_SUBJECT_TOKEN_TYPE2 = 'urn:ietf:params:oauth:token-type:jwt';
+
+/**
+ * Interface defining the JSON formatted response of a 3rd party executable
+ * used by the pluggable auth client.
+ */
+export interface ExecutableResponseJson {
+  /**
+   * The version of the JSON response. Only version 1 is currently supported.
+   * Always required.
+   */
+  version: number;
+  /**
+   * Whether the executable ran successfully. Always required.
+   */
+  success: boolean;
+  /**
+   * Epoch time for expiration of the token in seconds, required for successful
+   * responses.
+   */
+  expiration_time?: number;
+  /**
+   * Type of subject token in the response, currently supported values are:
+   * urn:ietf:params:oauth:token-type:saml2
+   * urn:ietf:params:oauth:token-type:id_token
+   * urn:ietf:params:oauth:token-type:jwt
+   */
+  token_type?: string;
+  /**
+   * Error code from executable, required when unsuccessful.
+   */
+  code?: string;
+  /**
+   * Error message from executable, required when unsuccessful.
+   */
+  message?: string;
+  /**
+   * ID token to be used as a subject token when token_type is id_token or jwt.
+   */
+  id_token?: string;
+  /**
+   * Response to be used as a subject token when token_type is saml2.
+   */
+  saml_response?: string;
+}
+
+/**
+ * Defines the response of a 3rd party executable run by the pluggable auth client.
+ */
+export class ExecutableResponse {
+  /**
+   * The version of the Executable response. Only version 1 is currently supported.
+   */
+  readonly version: number;
+  /**
+   * Whether the executable ran successfully.
+   */
+  readonly success: boolean;
+  /**
+   * Epoch time for expiration of the token in seconds.
+   */
+  readonly expirationTime?: number;
+  /**
+   * Type of subject token in the response, currently supported values are:
+   * urn:ietf:params:oauth:token-type:saml2
+   * urn:ietf:params:oauth:token-type:id_token
+   * urn:ietf:params:oauth:token-type:jwt
+   */
+  readonly tokenType?: string;
+  /**
+   * Error code from executable.
+   */
+  readonly errorCode?: string;
+  /**
+   * Error message from executable.
+   */
+  readonly errorMessage?: string;
+  /**
+   * Subject token from executable, format depends on tokenType.
+   */
+  readonly subjectToken?: string;
+
+  /**
+   * Instantiates an ExecutableResponse instance using the provided JSON object
+   * from the output of the executable.
+   * @param responseJson Response from a 3rd party executable, loaded from a
+   * run of the executable or a cached output file.
+   */
+  constructor(responseJson: ExecutableResponseJson) {
+    // Check that always required fields exist in the json response.
+    if (!responseJson.version) {
+      throw Error("Executable response must contain a 'version' field.");
```

Comment:
I think there would be value in creating typed errors, as it makes flow control easier.

```
try {

} catch (err) {
  if (err instanceof MissingVersion) {
     // do something.
  } else {
    throw err.
  }
}
```

We could consider introducing an `errors.ts` file if we find that we're using the same errors across files.

Reply:
Added typed errors for specific executableresponse validation issues